### ISSUE TITLE
Remote Comment capability

### ISSUE THREAD

**Issue by drequivalent** [Chocobozzz/PeerTube#224]
When I follow an account from Mastodon, I'm able to response only to new videos, that were uploaded since my server started federating with PeerTube.
I'd like to be able to comment on older videos, **no matter if my instance is federating with PeerTube at the moment or not.**

---

### Status:

Basically we are blocked by the fediverse support for intents. Typically remote 'follow' and 'share' intents are supported, but not yet the remote 'reply' intent¹. As stated in the issue this means the commenting account's instance has to federate first to be able to reply.

- [ ] Mastodon support for 'reply' intents via AP (pending https://github.com/tootsuite/mastodon/issues/916)
- [ ] Pleroma support for 'reply' intents via AP (confirmed unsupported after discussing on IRC)
- [ ] Hubzilla support for 'reply' intents via AP (?)
- [ ] postActiv support for 'reply' intents via AP (?)

Diaspora is not using ActivityPub at the moment, so until they do or until we federate with their network, we won't be able to dig any further. As a matter of fact here are the networks blocked until we support OStatus or they support AP:

- [ ] Diaspora support for 'reply' intents via OStatus (?)
- [ ] Friendica support for 'reply' intents via OStatus (?)
- [ ] GNU Social (ex StatusNet) support for 'reply' intents via OStatus (?)

¹: there is no standard in the AP realm for Intents. One is drafted here: https://akihikodaki.github.io/activity-intent/index | without a standard, it means the implementations are server-specific, hence the list of support per server.

**Comment by bnjbvr** [comment]
Correct me if I'm wrong, but I think ActivityPub doesn't allow you to do this in the way it works: it just emits metadata notifications for new content. For instance, you cannot get older toots on Mastodon after following somebody for the first time. (Maybe there's something to do at the standard specification level itself?)

**Comment by drequivalent** [comment]
Try to copy an URI of an old toot on an instance that does not federate with you. Mastodon will find this toot and let you respond.

**Comment by bnjbvr** [comment]
I see, makes sense.

**Comment by drequivalent** [comment]
It's kinda the same mechanism as with "Remote Follow", AFAIK.

**Comment by Chocobozzz** [comment]
I think this is a feature request on the Mastodon side :thinking:

**Comment by drequivalent** [comment]
Will Mastodon implement a "Remote comment" link on Peertube pages? I don't think so.

**Comment by Chocobozzz** [comment]
I'm sorry but I'm not sure to understand. I thought you wanted that on Mastodon, when you search a  PeerTube video with its URL you have a result (which is not the case for now).

**Comment by drequivalent** [comment]
That would be a poor man's version of what I wanted.

What's better is "Remote comment" link right above the comments section which then asks for your Fediverse account and then redirects you to your Mastodon or whatnot with your video's page open ready for response.

**Comment by drequivalent** [comment]
Any news on this one?

**Comment by rigelk** [comment]
Maybe I'm mistaken but Mastodon, for instance, only supports `follow` and `share` intents ([source](https://github.com/tootsuite/mastodon/blob/40e5d2303ba1edc51beae66cc15263675980106a/app/controllers/intents_controller.rb#L9-L13)).

There is https://github.com/tootsuite/mastodon/issues/916 tracking the implementation of such a feature.
I checked quickly on other fediverse software but couldn't find the feature either.

Any idea as how to implement this despite the support by others, @drequivalent ?

**Comment by ericblank** [comment]
Mastodon definitely shows me people's older posts from before I added them @bnjbvr

**Comment by johynpapin** [comment]
@leethax666 only on your own instance I think

**Comment by riking** [comment]
@Chocobozzz I think the issue is that PeerTube is not giving JSON/XML AP responses for video links. https://github.com/tootsuite/mastodon/blob/master/app/services/fetch_atom_service.rb#L31-L32

A link such as https://peertube.cpy.re/videos/watch/da2b08d4-a242-4170-b32a-4ec8cbdca701 should have this in the HTML: `<link rel=alternate type="application/atom+xml" href="...">` `<link rel=alternate type="application/activity+json" href="...">` (make sure to update these links on navigation! I want to write a browser extension)

And respond to matching `Accept:` headers with the OStatus or ActivityPub views of the video.

**Comment by Chocobozzz** [comment]
@riking Hello,

We give AP responses for video links: https://github.com/Chocobozzz/PeerTube/blob/develop/server/controllers/activitypub/client.ts#L44

Try 
```
$ curl -H "Accept: application/ld+json" https://peertube.cpy.re/videos/watch/5a6133b8-3e0c-40dc-b859-69d0540c3fe5
```

For the HTML `<link>` you're right! I created a [separate issue](https://github.com/Chocobozzz/PeerTube/issues/630) because as @rigelk said we are blocked by Angular.

**Comment by rigelk** [comment]
Now that Mastodon 2.5.0 implements remote reply/boost interactions through https://github.com/tootsuite/mastodon/pull/8202, at least we can redirect mastodon users to interact directly with the video.

**Comment by rigelk** [comment]
Implemented in 660d11e91e1643927028d2d6870a911f569b34d8.

**Comment by SerCom-KC** [comment]
Consider removing https://github.com/Chocobozzz/PeerTube/blob/28da43cf6606c052605a59662341075ba10666ee/client/src/app/videos/%2Bvideo-watch/comment/video-comments.component.html#L8 ?

**Comment by rigelk** [comment]
@SerCom-KC on my todo :wink: I will integrate part of its content with the modal.
